File: adapters/SkyrimPexStringTool/PexVisibleApiRegistry.cs
using System.Security.Cryptography;
using System.Text.Json;

internal sealed record PexVisibleArgumentRule(
    int Index,
    string SemanticRole,
    string Classification,
    string VisibilityBasis);

internal sealed record PexVisibleApiRule(
    string Callee,
    HashSet<string> OpcodeForms,
    IReadOnlyDictionary<int, PexVisibleArgumentRule> Arguments);

internal sealed class PexVisibleApiRegistry
{
    private const int SupportedSchemaVersion = 1;
    private static readonly HashSet<string> AllowedRootFields =
    [
        "schema_version",
        "game_id",
        "literal_policy",
        "unmatched_classification",
        "dynamic_argument_classification",
        "apis",
    ];
    private static readonly HashSet<string> AllowedApiFields =
    [
        "callee",
        "opcode_forms",
        "arguments",
        "evidence",
    ];
    private static readonly HashSet<string> AllowedArgumentFields =
    [
        "index",
        "semantic_role",
        "classification",
    ];
    private static readonly HashSet<string> AllowedOpcodes =
    [
        "CALLMETHOD",
        "CALLPARENT",
        "CALLSTATIC",
    ];

    private readonly Dictionary<string, PexVisibleApiRule> _apis;

    private PexVisibleApiRegistry(
        string gameId,
        string sourcePath,
        string sourceSha256,
        Dictionary<string, PexVisibleApiRule> apis)
    {
        GameId = gameId;
        SourcePath = sourcePath;
        SourceSha256 = sourceSha256;
        _apis = apis;
    }

    public string GameId { get; }
    public string SourcePath { get; }
    public string SourceSha256 { get; }

    public static PexVisibleApiRegistry? LoadForGame(
        string gameId,
        string? explicitRegistryPath = null)
    {
        if (!string.Equals(gameId, "fallout4", StringComparison.Ordinal))
        {
            return null;
        }
        var path = string.IsNullOrWhiteSpace(explicitRegistryPath)
            ? ResolveRegistryPath(gameId)
            : Path.GetFullPath(explicitRegistryPath);
        if (!File.Exists(path))
        {
            throw new FileNotFoundException($"PEX visible API registry does not exist: {path}");
        }
        return Load(path, gameId);
    }

    public bool TryGetArgument(
        string callee,
        string opcode,
        int semanticArgumentIndex,
        out PexVisibleArgumentRule? rule)
    {
        rule = null;
        if (!_apis.TryGetValue(callee, out var api)
            || !api.OpcodeForms.Contains(opcode)
            || !api.Arguments.TryGetValue(semanticArgumentIndex, out var found))
        {
            return false;
        }
        rule = found;
        return true;
    }

    private static PexVisibleApiRegistry Load(string path, string expectedGameId)
    {
        var bytes = File.ReadAllBytes(path);
        using var document = JsonDocument.Parse(bytes);
        var root = document.RootElement;
        RequireObject(root, "PEX visible API registry");
        RequireExactFields(root, AllowedRootFields, "PEX visible API registry");
        if (root.GetProperty("schema_version").GetInt32() != SupportedSchemaVersion)
        {
            throw new InvalidDataException("Unsupported PEX visible API registry schema_version.");
        }
        var gameId = RequireText(root, "game_id", "PEX visible API registry");
        if (!string.Equals(gameId, expectedGameId, StringComparison.Ordinal))
        {
            throw new InvalidDataException(
                $"PEX visible API registry game_id '{gameId}' does not match '{expectedGameId}'.");
        }
        if (!string.Equals(RequireText(root, "literal_policy", "PEX visible API registry"), "direct_only", StringComparison.Ordinal)
            || !string.Equals(RequireText(root, "unmatched_classification", "PEX visible API registry"), "manual_review", StringComparison.Ordinal)
            || !string.Equals(RequireText(root, "dynamic_argument_classification", "PEX visible API registry"), "manual_review", StringComparison.Ordinal))
        {
            throw new InvalidDataException("PEX visible API registry fallback policy must fail closed.");
        }

        var apisElement = root.GetProperty("apis");
        if (apisElement.ValueKind != JsonValueKind.Array || apisElement.GetArrayLength() == 0)
        {
            throw new InvalidDataException("PEX visible API registry apis must be a non-empty array.");
        }
        var apis = new Dictionary<string, PexVisibleApiRule>(StringComparer.OrdinalIgnoreCase);
        var classifications = new HashSet<string>(StringComparer.Ordinal);
        foreach (var apiElement in apisElement.EnumerateArray())
        {
            RequireObject(apiElement, "PEX visible API entry");
            RequireExactFields(apiElement, AllowedApiFields, "PEX visible API entry");
            var callee = RequireText(apiElement, "callee", "PEX visible API entry");
            var opcodeValues = ReadUniqueTextArray(apiElement.GetProperty("opcode_forms"), "opcode_forms");
            var opcodes = opcodeValues.ToHashSet(StringComparer.Ordinal);
            if (opcodes.Any(opcode => !AllowedOpcodes.Contains(opcode)))
            {
                throw new InvalidDataException($"PEX visible API '{callee}' has an unsupported opcode.");
            }
            var evidence = ReadUniqueTextArray(apiElement.GetProperty("evidence"), "evidence");
            var argumentElements = apiElement.GetProperty("arguments");
            if (argumentElements.ValueKind != JsonValueKind.Array || argumentElements.GetArrayLength() == 0)
            {
                throw new InvalidDataException($"PEX visible API '{callee}' requires argument rules.");
            }
            var arguments = new Dictionary<int, PexVisibleArgumentRule>();
            foreach (var argumentElement in argumentElements.EnumerateArray())
            {
                RequireObject(argumentElement, "PEX visible API argument");
                RequireExactFields(argumentElement, AllowedArgumentFields, "PEX visible API argument");
                var index = argumentElement.GetProperty("index").GetInt32();
                if (index < 0)
                {
                    throw new InvalidDataException($"PEX visible API '{callee}' has a negative argument index.");
                }
                var role = RequireText(argumentElement, "semantic_role", "PEX visible API argument");
                var classification = RequireText(argumentElement, "classification", "PEX visible API argument");
                if (classification is not ("visible" or "protected"))
                {
                    throw new InvalidDataException($"PEX visible API '{callee}' has an unsafe classification.");
                }
                classifications.Add(classification);
                var basis = $"registry:{Path.GetFileName(path)}#{callee}[{index}]/{evidence[0]}";
                if (!arguments.TryAdd(index, new PexVisibleArgumentRule(index, role, classification, basis)))
                {
                    throw new InvalidDataException($"PEX visible API '{callee}' repeats argument index {index}.");
                }
            }
            if (!apis.TryAdd(callee, new PexVisibleApiRule(callee, opcodes, arguments)))
            {
                throw new InvalidDataException($"Duplicate PEX visible API callee '{callee}'.");
            }
        }
        if (!classifications.SetEquals(["visible", "protected"]))
        {
            throw new InvalidDataException("PEX visible API registry must contain visible and protected rules.");
        }
        return new PexVisibleApiRegistry(
            gameId,
            path,
            Convert.ToHexString(SHA256.HashData(bytes)),
            apis);
    }

    private static string ResolveRegistryPath(string gameId)
    {
        foreach (var start in new[] { AppContext.BaseDirectory, Directory.GetCurrentDirectory() })
        {
            var current = new DirectoryInfo(Path.GetFullPath(start));
            for (var depth = 0; current is not null && depth < 12; depth++, current = current.Parent)
            {
                var candidate = Path.Combine(current.FullName, "config", "pex_visible_apis", $"{gameId}.json");
                if (File.Exists(candidate))
                {
                    return Path.GetFullPath(candidate);
                }
            }
        }
        throw new FileNotFoundException(
            $"Could not locate the versioned PEX visible API registry for {gameId}.");
    }

    private static void RequireObject(JsonElement element, string label)
    {
        if (element.ValueKind != JsonValueKind.Object)
        {
            throw new InvalidDataException($"{label} must be an object.");
        }
    }

    private static void RequireExactFields(JsonElement element, HashSet<string> allowed, string label)
    {
        var actual = element.EnumerateObject().Select(property => property.Name).ToHashSet(StringComparer.Ordinal);
        if (!actual.SetEquals(allowed))
        {
            throw new InvalidDataException($"{label} fields do not match the supported schema.");
        }
    }

    private static string RequireText(JsonElement parent, string property, string label)
    {
        var element = parent.GetProperty(property);
        if (element.ValueKind != JsonValueKind.String || string.IsNullOrWhiteSpace(element.GetString()))
        {
            throw new InvalidDataException($"{label}.{property} must be non-empty text.");
        }
        return element.GetString()!.Trim();
    }

    private static IReadOnlyList<string> ReadUniqueTextArray(JsonElement element, string label)
    {
        if (element.ValueKind != JsonValueKind.Array || element.GetArrayLength() == 0)
        {
            throw new InvalidDataException($"PEX visible API {label} must be a non-empty array.");
        }
        var values = new List<string>();
        var seen = new HashSet<string>(StringComparer.Ordinal);
        foreach (var item in element.EnumerateArray())
        {
            if (item.ValueKind != JsonValueKind.String || string.IsNullOrWhiteSpace(item.GetString()))
            {
                throw new InvalidDataException($"PEX visible API {label} contains an empty value.");
            }
            var value = item.GetString()!.Trim();
            if (!seen.Add(value))
            {
                throw new InvalidDataException($"PEX visible API {label} contains a duplicate value.");
            }
            values.Add(value);
        }
        return values;
    }
}

File: scripts/pex_visible_api_registry.py
"""Load the fixture-backed PEX call-site visibility registry."""

from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any


SCHEMA_VERSION = 1
CALL_OPCODES = frozenset({"CALLMETHOD", "CALLPARENT", "CALLSTATIC"})
ARGUMENT_CLASSIFICATIONS = frozenset({"visible", "protected"})
FALLBACK_CLASSIFICATION = "manual_review"
_CALLEE = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*(?:\.[A-Za-z_][A-Za-z0-9_]*)+$")
_SEMANTIC_ROLE = re.compile(r"^[a-z][a-z0-9_]*$")


def _require_object(value: Any, label: str) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise ValueError(f"{label} must be an object")
    return value


def _require_exact_fields(
    value: dict[str, Any],
    expected: frozenset[str],
    label: str,
) -> None:
    missing = sorted(expected - value.keys())
    unknown = sorted(value.keys() - expected)
    if missing or unknown:
        details: list[str] = []
        if missing:
            details.append(f"missing {', '.join(missing)}")
        if unknown:
            details.append(f"unknown {', '.join(unknown)}")
        raise ValueError(f"{label} fields are invalid: {'; '.join(details)}")


def _non_empty_strings(value: Any, label: str) -> list[str]:
    if not isinstance(value, list) or not value:
        raise ValueError(f"{label} must be a non-empty array")
    if any(not isinstance(item, str) or not item.strip() for item in value):
        raise ValueError(f"{label} entries must be non-empty strings")
    normalized = [item.strip() for item in value]
    if len({item.casefold() for item in normalized}) != len(normalized):
        raise ValueError(f"{label} contains duplicate entries")
    return normalized


def load_pex_visible_api_registry(
    path: Path,
    *,
    expected_game_id: str | None = None,
) -> dict[str, Any]:
    try:
        payload = json.loads(path.read_text(encoding="utf-8-sig"))
    except json.JSONDecodeError as exc:
        raise ValueError(f"PEX visible API registry is invalid JSON: {path}") from exc
    root = _require_object(payload, "PEX visible API registry")
    _require_exact_fields(
        root,
        frozenset(
            {
                "schema_version",
                "game_id",
                "literal_policy",
                "unmatched_classification",
                "dynamic_argument_classification",
                "apis",
            }
        ),
        "PEX visible API registry",
    )
    if root["schema_version"] != SCHEMA_VERSION:
        raise ValueError(
            f"unsupported PEX visible API registry schema_version={root['schema_version']}"
        )
    game_id = root["game_id"]
    if not isinstance(game_id, str) or not game_id.strip():
        raise ValueError("PEX visible API registry game_id must be non-empty text")
    if expected_game_id is not None and game_id != expected_game_id:
        raise ValueError(
            f"PEX visible API registry game_id {game_id!r} does not match {expected_game_id!r}"
        )
    if root["literal_policy"] != "direct_only":
        raise ValueError("PEX visible API registry literal_policy must be direct_only")
    for field in ("unmatched_classification", "dynamic_argument_classification"):
        if root[field] != FALLBACK_CLASSIFICATION:
            raise ValueError(f"PEX visible API registry {field} must be manual_review")

    apis = root["apis"]
    if not isinstance(apis, list) or not apis:
        raise ValueError("PEX visible API registry apis must be a non-empty array")
    seen_callees: set[str] = set()
    classifications: set[str] = set()
    for api_index, raw_api in enumerate(apis):
        label = f"apis[{api_index}]"
        api = _require_object(raw_api, label)
        _require_exact_fields(
            api,
            frozenset({"callee", "opcode_forms", "arguments", "evidence"}),
            label,
        )
        callee = api["callee"]
        if not isinstance(callee, str) or _CALLEE.fullmatch(callee) is None:
            raise ValueError(f"{label}.callee is invalid: {callee!r}")
        callee_key = callee.casefold()
        if callee_key in seen_callees:
            raise ValueError(f"duplicate PEX visible API callee: {callee}")
        seen_callees.add(callee_key)

        opcodes = _non_empty_strings(api["opcode_forms"], f"{label}.opcode_forms")
        unsupported_opcodes = sorted(set(opcodes) - CALL_OPCODES)
        if unsupported_opcodes:
            raise ValueError(
                f"{label}.opcode_forms contains unsupported values: "
                f"{', '.join(unsupported_opcodes)}"
            )
        _non_empty_strings(api["evidence"], f"{label}.evidence")

        arguments = api["arguments"]
        if not isinstance(arguments, list) or not arguments:
            raise ValueError(f"{label}.arguments must be a non-empty array")
        seen_indexes: set[int] = set()
        for argument_position, raw_argument in enumerate(arguments):
            argument_label = f"{label}.arguments[{argument_position}]"
            argument = _require_object(raw_argument, argument_label)
            _require_exact_fields(
                argument,
                frozenset({"index", "semantic_role", "classification"}),
                argument_label,
            )
            index = argument["index"]
            if isinstance(index, bool) or not isinstance(index, int) or index < 0:
                raise ValueError(f"{argument_label}.index must be a non-negative integer")
            if index in seen_indexes:
                raise ValueError(f"{label}.arguments contains duplicate index {index}")
            seen_indexes.add(index)
            role = argument["semantic_role"]
            if not isinstance(role, str) or _SEMANTIC_ROLE.fullmatch(role) is None:
                raise ValueError(f"{argument_label}.semantic_role is invalid: {role!r}")
            classification = argument["classification"]
            if classification not in ARGUMENT_CLASSIFICATIONS:
                raise ValueError(
                    f"{argument_label}.classification must be visible or protected"
                )
            classifications.add(classification)

    missing_classes = ARGUMENT_CLASSIFICATIONS - classifications
    if missing_classes:
        raise ValueError(
            "PEX visible API registry must include fixture-backed visible and protected "
            f"arguments; missing {', '.join(sorted(missing_classes))}"
        )
    return root
